### PR TITLE
[large scale]Get next job id from gcs instead of redis - python part

### LINKED ISSUE #15358: Move java lang worker off of redis
TODO: Scope this

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
## Why are these changes needed?
Follow up of https://github.com/ray-project/ray/pull/16190, we change the python part this time.

## Related issue number

Close #15358 and #13984
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by WangTaoTheTonic** [comment]
IDK if we can directly move the initialization of `ray.state` up to the position before core worker. Hope somebody can tell me the impact. :-)

File: python/ray/tests/test_global_state.py [hunk 1]
```diff
@@ -357,6 +357,10 @@ def self_ip_is_set():
     global_state_accessor.disconnect()
 
 
+def test_next_job_id(ray_start_regular):
+    assert ray.state.next_job_id().int() + 1 == ray.state.next_job_id().int()
```
Comment (kfstorm) [inline python/ray/tests/test_global_state.py:361]: This looks tricky to me. How about

```
job_id_1 = ray.state.next_job_id()
job_id_2 = ray.state.next_job_id()
assert job_id_1.int() + 1 == job_id_2.int()
```

File: python/ray/state.py [hunk 2]
```diff
@@ -771,6 +781,15 @@ def jobs():
     return state.job_table()
 
 
+def next_job_id():
+    """Get next job id from GCS.
+
+    Returns:
+        Information about the node in the cluster.
```
Comment (kfstorm) [inline python/ray/state.py:788]: Wrong documentation about the return value.

File: python/ray/includes/global_state_accessor.pxi [hunk 3]
```diff
@@ -43,6 +43,12 @@ cdef class GlobalStateAccessor:
             result = self.inner.get().GetAllJobInfo()
         return result
 
+    def get_next_job_id(self):
+        cdef CJobID cjob_id
+        with nogil:
+            cjob_id = self.inner.get().GetNextJobID()
+        return cjob_id.ToInt()
```
Comment (kfstorm) [inline python/ray/includes/global_state_accessor.pxi:50]: Why do we need `.ToInt()` here?

File: python/ray/state.py [hunk 4]
```diff
@@ -277,6 +277,16 @@ def job_table(self):
 
         return results
 
+    def next_job_id(self):
```
Comment (kfstorm) [inline python/ray/state.py:280]: Why two `next_job_id` definitions?

**Comment by kfstorm** [comment]
> IDK if we can directly move the initialization of `ray.state` up to the position before core worker. Hope somebody can tell me the impact. :-)

As long as we keep the output of `ray.init()` clean, I'm OK to move it up a bit.

A long-term solution would be to only print error logs when `RayLog` is not initialized yet.

File: python/ray/tests/test_global_state.py (hunk 1 above)
Comment (WangTaoTheTonic) [reply python/ray/tests/test_global_state.py:361]: No problem. I wrote like this in first version. :-)

File: python/ray/includes/global_state_accessor.pxi (hunk 3 above)
Comment (WangTaoTheTonic) [reply python/ray/includes/global_state_accessor.pxi:50]: The python level cannot understand CJobID.

File: python/ray/state.py (hunk 4 above)
Comment (WangTaoTheTonic) [reply python/ray/state.py:280]: One is in state, another is in `GlobalState` used by state.

File: python/ray/includes/global_state_accessor.pxi (hunk 3 above)
Comment (kfstorm) [reply python/ray/includes/global_state_accessor.pxi:50]: Then why not return a Python JobID here?

File: python/ray/state.py (hunk 4 above)
Comment (kfstorm) [reply python/ray/state.py:280]: Not all methods in state.py have two definitions. Do you know when we need two and when we need only one?

File: python/ray/includes/global_state_accessor.pxi (hunk 3 above)
Comment (WangTaoTheTonic) [reply python/ray/includes/global_state_accessor.pxi:50]: The `inner` variable is a CGlobalStateAccessor pointer which **cimport** from `global_state_accessor.h`, so it cannot return Python JobID.

File: python/ray/state.py (hunk 4 above)
Comment (WangTaoTheTonic) [reply python/ray/state.py:280]: All methods except those related with placement group are first defined in `GlobalState` then referred in state.py. I believe the state.py is a (global) wrapper of `GlobalState` to separate the interfaces and implementations.
For placement group part I'd rather think it's a missing part.

**Review by kfstorm — APPROVED** [review]

File: python/ray/includes/global_state_accessor.pxi (hunk 3 above)
Comment (wuisawesome) [reply python/ray/includes/global_state_accessor.pxi:50]: Sorry, I don't think I understand. Can we do `JobID(cjob_id.Binary())` here?

**Review by wuisawesome — APPROVED** [review]

File: python/ray/includes/global_state_accessor.pxi (hunk 3 above)
Comment (WangTaoTheTonic) [reply python/ray/includes/global_state_accessor.pxi:50]: Yes maybe we can. But I think the better way is to keep the current change and do serialize part in `GlobalState` in state.py, like other interfaces.
BTW I'd rather use `.ToInt` as the origin value is from an integer type value.

**Pull request merged**